wallet: Add missing cs_wallet/cs_KeyStore locks to wallet

practicalswift opened the pull request:
Add missing wallet locks:

* Calling the function `GetConflicts(...)` requires holding the mutex `cs_wallet`
* Calling the function `IsSpent(...)` requires holding the mutex `cs_wallet`
* Accessing the variables `mapKeys` and `mapCryptedKeys` requires holding the mutex `cs_KeyStore`
* Accessing the variable `nTimeFirstKey` requires holding the mutex `cs_wallet`
* Accessing the variable `mapWallet` requires holding the mutex `cs_wallet`
* Accessing the variable `nTimeFirstKey` requires holding the mutex `cs_wallet`

promag commented on src/wallet/wallet.cpp:
```diff
@@ -3953,6 +3960,7 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
             for (const CWalletTx& wtxOld : vWtx)
             {
                 uint256 hash = wtxOld.GetHash();
+                LOCK(walletInstance->cs_wallet);
```
Either:
 - lock once before the loop;
 - loop first with the lock to `mapWallet.find(hash)` for all `vWtx` and then loop again without the lock to `walletdb.WriteTx()`;
 - lock only the mapWallet.find(hash).

practicalswift commented:
@promag Thanks for reviewing! Feedback addressed. Looks good? :-)

promag commented on src/wallet/wallet.cpp:
```diff
@@ -3950,6 +3957,7 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
         {
             CWalletDB walletdb(*walletInstance->dbw);
 
+            LOCK(walletInstance->cs_wallet);
```
This should be before the line above. Anyway, there is no need to lock this wallet since it's a fresh instance, unknown to the remaining system. Unless you are adding the lock because of other asserts and if so please commit them to justify this change?

practicalswift replied:
Now moved before the line above. Without that lock we'll trigger Clang thread safety analysis warnings when #11634 is merged due to:
```
src/wallet/wallet.h:    std::map<uint256, CWalletTx> mapWallet GUARDED_BY(cs_wallet);
```

As agreed upon in https://github.com/bitcoin/bitcoin/pull/11226#issuecomment-341496939 all guard/lock annotations are added in #11634 and all extra locking is submitted as separate PR:s (such as this one).

The former are compile-time only whereas the latter might change run-time behaviour (and thus needs extra scrunity!).

promag replied:
> As agreed upon in #11226 (comment) all guard/lock annotations are added in #11634 and all extra locking is submitted as separate PR:s (such as this one)

I think you swapped the PR numbers?

practicalswift replied:
Yes, you're right - the annotations are added in #11226 :-)

promag commented:
utACK 007fcbf.

practicalswift commented:
Added another commit with two additional missing locks:

* calling function `IsSpent` requires holding mutex `pwallet->cs_wallet` exclusively
* writing variable `nWalletVersion`, `nWalletMaxVersion`, `nOrderPosNext` and `nTimeFirstKey` require holding mutex `cs_wallet`

practicalswift commented:
@promag Would you mind re-reviewing? :-)

TheBlueMatt commented on src/wallet/wallet.h:
```diff
@@ -789,18 +789,21 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
-        nWalletVersion = FEATURE_BASE;
-        nWalletMaxVersion = FEATURE_BASE;
+        {
+            LOCK(cs_wallet);
```
Same here. Just add a cs_wallet lock for the whole function.

practicalswift commented:
@TheBlueMatt Thanks for reviewing! Feedback addressed. Please re-review :-)

practicalswift committed:
wallet: Add missing locks

practicalswift commented:
Fixed build issue. Please re-review :-)

MarcoFalke commented:
Given that https://github.com/bitcoin/bitcoin/pull/11226/files#diff-12635a58447c65585f51d32b7e04075bR857 is now closed, wouldn't it make sense to add the clang annotations within this commit?

practicalswift commented:
@MarcoFalke @TheBlueMatt @promag Thanks for reviewing. I've now addressed the feedback and added corresponding `GUARDED_BY`/`EXCLUSIVE_LOCKS_REQUIRED` annotations. Please re-review :-)

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -3930,6 +3938,7 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
     bool fFirstRun = true;
     CWallet *walletInstance = new CWallet(name, CWalletDBWrapper::Create(path));
+    LOCK(walletInstance->cs_wallet);
     DBErrors nLoadWalletRet = walletInstance->LoadWallet(fFirstRun);
```
Note that `LoadWallet` takes `LOCK2(cs_main, cs_wallet)`, so taking `cs_wallet` before `cs_main` leads to a deadlock in case another thread takes `cs_main` and then `cs_wallet`, no?

practicalswift replied:
Thanks. Now fixed. See https://github.com/bitcoin/bitcoin/pull/11634#discussion_r173839102.

practicalswift commented on src/wallet/wallet.cpp:
```diff
@@ -3991,8 +3996,11 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
-        while (pindexRescan && walletInstance->nTimeFirstKey && (pindexRescan->GetBlockTime() < (walletInstance->nTimeFirstKey - TIMESTAMP_WINDOW))) {
-            pindexRescan = chainActive.Next(pindexRescan);
+        {
+            LOCK(walletInstance->cs_wallet);
```
@TheBlueMatt Moving it up will make the scope of the `walletInstance->cs_wallet` lock cover the `ScanForWalletTransactions` call. `ScanForWalletTransactions` calls `GuessVerificationProgress` which locks `cs_main` giving us the lock order:
* 1. `cs_wallet`
* 2. `cs_main`

This is a potential deadlock given that the opposite lock order is used extensively:

```
$ git grep "LOCK2(cs_main,.*cs_wallet);" | wc -l
89
```

Please advice :-)

promag replied:
A solution to that is to also lock cs_main.. but meh

practicalswift commented:
Please re-review :-)

practicalswift commented:
Rebased!

Having this merged would have catched this locking incident: https://github.com/bitcoin/bitcoin/pull/12565/files#r171235800

Reviews welcome! Perhaps @promag and @Sjors could take a look? :-)

promag commented on src/wallet/wallet.cpp:
```diff
@@ -1837,7 +1837,10 @@ std::set<uint256> CWalletTx::GetConflicts() const
         uint256 myHash = GetHash();
-        result = pwallet->GetConflicts(myHash);
+        {
+            LOCK(pwallet->cs_wallet);
```
I believe this lock is not necessary, all callers have the lock already.

Sjors commented:
Concept ACK: anything that prevents me from making mistakes :-)

practicalswift commented:
@promag Thanks for reviewing. Feedback addressed. Please review :-)

sipa commented:
Concept ACK

practicalswift commented on src/wallet/wallet.cpp:
```diff
@@ -1935,6 +1935,7 @@ CAmount CWalletTx::GetAvailableCredit(bool fUseCache) const
 
     CAmount nCredit = 0;
     uint256 hashTx = GetHash();
+    LOCK(pwallet->cs_wallet);
```
Is `pcoin->pwallet->cs_wallet == pwallet->cs_wallet` guaranteed?

promag replied:
Yes, otherwise we have a big problem I think 😛

practicalswift replied:
Unfortunately the Clang's thread-safety analysis does not seem to understand that :-(  It seems to think the locks can be different.

practicalswift replied:
@promag I think I'm unable to resolve this. Are you able to make this work with `EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)` on `CWalletTx::GetAvailableCredit` when compiling with Clang's thread-safety analysis enabled?

MarcoFalke replied:
Agree with @promag . We wouldn't want to add LOCKs just to please a static analyser that isn't smart enough to figure out the lock is already taken.

practicalswift commented on src/wallet/wallet.cpp:
```diff
@@ -1978,6 +1979,7 @@ CAmount CWalletTx::GetAvailableWatchOnlyCredit(const bool fUseCache) const
         return nAvailableWatchCreditCached;
 
     CAmount nCredit = 0;
+    LOCK(pwallet->cs_wallet);
```
@promag See above. In theory you're right, but I've been unable to convince Clang's thread-safety analysis that so is the case in practice. See comment above. I might need some help here :-)

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -708,7 +708,7 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     std::map<CKeyID, int64_t> m_pool_key_to_index;
 
-    int64_t nTimeFirstKey = 0;
+    int64_t nTimeFirstKey GUARDED_BY(cs_wallet);
```
Why remove the initialization?

practicalswift replied:
Ouch, that was not the intention. Thanks for noticing. Please re-review.

practicalswift commented:
Rebased!

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -319,7 +319,7 @@ bool CWallet::LoadCryptedKey(const CPubKey &vchPubKey, const std::vector<unsigne
  * are added to the wallet, with the oldest key creation time.
  */
-void CWallet::UpdateTimeFirstKey(int64_t nCreateTime)
+void CWallet::UpdateTimeFirstKey(int64_t nCreateTime) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
```
All those are redundant to the header

practicalswift commented:
@MarcoFalke Thanks for the review. I've now reworked this PR and moved annotations to the `.h` files. Could you please re-review? :-)

MarcoFalke commented:
`GetConflicts` has still an annotation in the cpp file?

practicalswift commented:
@MarcoFalke 

Applying …

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index a74efb919..f2926fd74 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1865,7 +1865,7 @@ bool CWalletTx::RelayWalletTransaction(CConnman* connman)
     return false;
 }

-std::set<uint256> CWalletTx::GetConflicts() const EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+std::set<uint256> CWalletTx::GetConflicts() const
 {
     std::set<uint256> result;
     if (pwallet != nullptr)
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index d1ffcfbdc..fa4ca5ccb 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -493,7 +493,7 @@ public:
     /** Pass this transaction to the mempool. Fails if absolute fee exceeds absurd fee. */
     bool AcceptToMemoryPool(const CAmount& nAbsurdFee, CValidationState& state);

-    std::set<uint256> GetConflicts() const;
+    std::set<uint256> GetConflicts() const EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet);
 };

 class COutput
```

… results in …

```
In file included from wallet/feebumper.cpp:6:
In file included from ./wallet/coincontrol.h:11:
./wallet/wallet.h:496:76: error: member access into incomplete type 'const CWallet'
    std::set<uint256> GetConflicts() const EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet);
                                                                           ^
```

OTOH, applying only …

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index a74efb919..f2926fd74 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1865,7 +1865,7 @@ bool CWalletTx::RelayWalletTransaction(CConnman* connman)
     return false;
 }

-std::set<uint256> CWalletTx::GetConflicts() const EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+std::set<uint256> CWalletTx::GetConflicts() const
 {
     std::set<uint256> result;
     if (pwallet != nullptr)
```

… results in …

```
wallet/wallet.cpp:1874:27: error: calling function 'GetConflicts' requires holding mutex 'pwallet->cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
        result = pwallet->GetConflicts(myHash);
                          ^
```

practicalswift commented:
@MarcoFalke Another alternative could be to add `NO_THREAD_SAFETY_ANALYSIS` here to disable the analysis locally.

What would you suggest as the recommended way to proceed?

MarcoFalke commented:
@practicalswift I like your latest suggestion (`NO_THREAD_SAFETY_ANALYSIS`). Make sure to include a comment to explain this is only temporary. Also, explain that this is safe to do, since we have a run-time `AssertLockHeld` in place.

practicalswift commented:
@MarcoFalke Good point! Added and documented `NO_THREAD_SAFETY_ANALYSIS`. Please review :-)

practicalswift commented:
Rebased!

practicalswift commented:
Rebased!

practicalswift commented:
Rebased!

practicalswift commented:
Rebased!

practicalswift commented:
Rebase number eight performed! :-)

practicalswift commented:
Rebased!

practicalswift commented:
@TheBlueMatt @promag Feedback addressed (added two commits to keep changes easy to review). 

@MarcoFalke @sipa @Sjors Please re-review :-)

MarcoFalke commented:
@practicalswift Could squash into two commits? First one is adding the `LOCK`s, the second one the annotations?

practicalswift committed:
wallet: Add Clang thread safety analysis annotations

practicalswift commented:
@MarcoFalke Done! Please re-review :-)

MarcoFalke commented:
utACK 75cb9c068570ec433f6ebc966f0771caa92efe81

practicalswift commented:
Rebased!

@MarcoFalke @promag  - please re-review your utACK:s :-)

practicalswift committed:
Add GUARDED_BY(cs_wallet) for encrypted_batch, nWalletMaxVersion, m_max_keypool_index and nOrderPosNext

* AddKeyPubKeyWithDB(...) reads encrypted_batch which potentially races with write in the same method.
* IncOrderPosNext(...) reads nOrderPosNext which potentially races with write in BlockDisconnected(...).
* LoadKeyPool(...) reads m_max_keypool_index which potentially races with write in BlockDisconnected(...).
* LoadMinVersion(...) reads nWalletMaxVersion which potentially races with write in BlockDisconnected(...).

practicalswift committed:
Add GUARDED_BY(cs_wallet) for setExternalKeyPool, mapKeyMetadata, m_script_metadata and setLockedCoins

practicalswift commented:
Added `GUARDED_BY(cs_wallet)` for `setExternalKeyPool`, `mapKeyMetadata`, `m_script_metadata` and `setLockedCoins`.

Rationale:

```
$ git grep -E 'AssertLockHeld.*(setExternalKeyPool|mapKeyMetadata|m_script_metadata|setLockedCoins)' | sort | uniq -c
      4 src/wallet/wallet.cpp:    AssertLockHeld(cs_wallet); // mapKeyMetadata
      1 src/wallet/wallet.cpp:    AssertLockHeld(cs_wallet); // m_script_metadata
      1 src/wallet/wallet.cpp:    AssertLockHeld(cs_wallet); // setExternalKeyPool
      5 src/wallet/wallet.cpp:    AssertLockHeld(cs_wallet); // setLockedCoins
```

practicalswift commented:
@MarcoFalke Would you mind reviewing the updated version? :-)

MarcoFalke commented:
utACK 69e7ee2dd8173597e766262fd9a8caae569ddf5e

promag commented on src/wallet/wallet.cpp:
```diff
@@ -4093,7 +4093,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
     walletInstance->TopUpKeyPool();
 
-    LOCK(cs_main);
+    LOCK2(cs_main, walletInstance->cs_wallet);
```
No annotation requires this lock right?

practicalswift replied:
Removing that locks results in the following:

```
wallet/wallet.cpp:4104:28: error: calling function 'FindForkInGlobalIndex' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            pindexRescan = FindForkInGlobalIndex(chainActive, locator);
                           ^
wallet/wallet.cpp:4131:48: error: reading variable 'nTimeFirstKey' requires holding mutex 'walletInstance->cs_wallet' [-Werror,-Wthread-safety-analysis]
        while (pindexRescan && walletInstance->nTimeFirstKey && (pindexRescan->GetBlockTime() < (walletInstance->nTimeFirstKey - TIMESTAMP_WINDOW))) {
                                               ^
wallet/wallet.cpp:4131:114: error: reading variable 'nTimeFirstKey' requires holding mutex 'walletInstance->cs_wallet' [-Werror,-Wthread-safety-analysis]
        while (pindexRescan && walletInstance->nTimeFirstKey && (pindexRescan->GetBlockTime() < (walletInstance->nTimeFirstKey - TIMESTAMP_WINDOW))) {
                                                                                                                 ^
wallet/wallet.cpp:4156:77: error: reading variable 'mapWallet' requires holding mutex 'walletInstance->cs_wallet' [-Werror,-Wthread-safety-analysis]
                std::map<uint256, CWalletTx>::iterator mi = walletInstance->mapWallet.find(hash);
                                                                            ^
wallet/wallet.cpp:4157:43: error: reading variable 'mapWallet' requires holding mutex 'walletInstance->cs_wallet' [-Werror,-Wthread-safety-analysis]
                if (mi != walletInstance->mapWallet.end())
                                          ^
wallet/wallet.cpp:4181:90: error: calling function 'GetKeyPoolSize' requires holding mutex 'walletInstance->cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
        walletInstance->WalletLogPrintf("setKeyPool.size() = %u\n",      walletInstance->GetKeyPoolSize());
                                                                                         ^
wallet/wallet.cpp:4182:90: error: reading variable 'mapWallet' requires holding mutex 'walletInstance->cs_wallet' [-Werror,-Wthread-safety-analysis]
        walletInstance->WalletLogPrintf("mapWallet.size() = %u\n",       walletInstance->mapWallet.size());
                                                                                         ^
```

promag replied:
Rigth, the new annotations..

It could make sense to avoid calling `uiInterface.LoadWallet()` with the lock held.

practicalswift replied:
@promag Ah, now I follow. Suggested fix below. What do you think?

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 29014790e..2deeb9c42 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4093,7 +4093,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
     // Try to top up keypool. No-op if the wallet is locked.
     walletInstance->TopUpKeyPool();

-    LOCK2(cs_main, walletInstance->cs_wallet);
+    LOCK(cs_main);

     CBlockIndex *pindexRescan = chainActive.Genesis();
     if (!gArgs.GetBoolArg("-rescan", false))
@@ -4128,6 +4128,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,

         // No need to read and scan block if block was created before
         // our wallet birthday (as adjusted for block time variability)
+        LOCK(walletInstance->cs_wallet);
         while (pindexRescan && walletInstance->nTimeFirstKey && (pindexRescan->GetBlockTime() < (walletInstance->nTimeFirstKey - TIMESTAMP_WINDOW))) {
             pindexRescan = chainActive.Next(pindexRescan);
         }
@@ -4178,6 +4179,7 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
     walletInstance->SetBroadcastTransactions(gArgs.GetBoolArg("-walletbroadcast", DEFAULT_WALLETBROADCAST));

     {
+        LOCK(walletInstance->cs_wallet);
         walletInstance->WalletLogPrintf("setKeyPool.size() = %u\n",      walletInstance->GetKeyPoolSize());
         walletInstance->WalletLogPrintf("mapWallet.size() = %u\n",       walletInstance->mapWallet.size());
         walletInstance->WalletLogPrintf("mapAddressBook.size() = %u\n",  walletInstance->mapAddressBook.size());
```

promag replied:
I think that currently there is no harm in calling ` uiInterface.LoadWallet() with the lock held but if it's not necessary then I wouldn't change that.

promag reviewed (commented):
utACK 69e7ee2.
